format/puffin-spec.md

```diff
@@ -123,6 +123,44 @@ The blob metadata for this blob may include following properties:
 
 - `ndv`: estimate of number of distinct values, derived from the sketch.
 
+#### `delete-vector-v1` blob type
+
+A serialized delete vector that represents the positions of rows in a file that
+are deleted.  A set bit at position P indicates that the row at position P is
+deleted.
+
+The bitmap supports positive 64-bit positions, but is optimized for cases where
+most positions fit in 32 bits by using a collection of 32-bit Roaring bitmaps.
+64-bit positions are divided into a 32-bit "key" using the most significant 4
+bytes and a 32-bit position using the least significant 4 bytes. For each key
+in the set of positions, a 32-bit Roaring bitmap is maintained to store a set
+of 32-bit positions for that key.
+
+To test whether a certain position is set, its most significant 4 bytes (the
+key) are used to find a 32-bit bitmap and the least significant 4 bytes are
+tested for inclusion in the bitmap. If a bitmap is not found for the key, then
+it is not set.
+
+The serialized blob starts with a 4-byte magic sequence, `D1D33964` (1681511377
+stored as 4 bytes, little-endian). Following the magic bytes is the serialized
+collection of bitmaps. The collection is stored using the Roaring bitmap
+["portable" format][roaring-bitmap-portable-serialization]. This representation
+consists of:
+
+* The number of 32-bit Roaring bitmaps, serialized as 8 bytes, little-endian
+* For each 32-bit Roaring bitmap, ordered by unsigned comparison of the 32-bit keys:
+    - The key stored as 4 bytes, little-endian
+    - A [32-bit Roaring bitmap][roaring-bitmap-general-layout]
+
+The blob metadata must include the following properties:
+
+* `referenced-data-file`: location of the data file the delete vector applies to
```

Comment:
I wasn't quite aware that we gonna put the deletion vectors in the puffin files. So now they gonna be mandatory for readers, right?
How this compares to alternatives considered? (i believe following delta example we could have deletion vectors inline in manifest list)

Reply:
an interesting discussion on the same : https://docs.google.com/document/d/18Bqhr-vnzFfQk1S4AgRISkA_5_m5m32Nnc2Cw0zn2XM/edit?disco=AAABUXU4vh4

Reply:
Technically a reader just needs to be able to read the vector since the addressing is all within the delete file manifest still.

My TLDR on the reasoning from Anton was, we make less files, but we still get the ability to individually address and attribute blobs. I would check the proposal @singhpk234 linked to

Reply:
thanks @singhpk234 @RussellSpitzer . So these Puffin won't be registered on the table-level, like the table stats files are, right?

Reply:
@findepi, Puffin is used to make the files self-describing, but it is not necessary to read the Puffin footer in order to apply the delete vector. We are basically just using it as a container for DVs.

Reply:
so the idea is that manifest will link to portions of the file (by path+offset+length)?
but the referenced-data-file stays important property.

i am concerned that, if it's self-describing but portions of information isn't used at execution, this leaves space for bugs to creep in. the self-descibing but unused portion may end up containing incorrect information, without anyone noticing (either for files produced by Iceberg itself, or by other implementors/query engines, who use Iceberg project for compatibility test purposes).

Reply:
PR #11302 has a sample implementation. The magic number and CRC in the blob ensures we read a correct byte sequence. The Puffin footer will be helpful to debug potential issues or during table maintenance. It enables us to store the referenced data file next to the blob in addition to storing it in the metadata.